File: ili9341.py
from machine import Pin, SPI
from time import sleep_ms
import framebuf

# ILI9341 Commands
ILI9341_SWRESET = 0x01
ILI9341_SLPOUT = 0x11
ILI9341_DISPON = 0x29
ILI9341_CASET = 0x2A
ILI9341_PASET = 0x2B
ILI9341_RAMWR = 0x2C
ILI9341_MADCTL = 0x36
ILI9341_COLMOD = 0x3A

# Memory Access Control (MADCTL) Bits
MADCTL_MY = 0x80  # Mirror Y
MADCTL_MX = 0x40  # Mirror X
MADCTL_MV = 0x20  # Memory Data Access Control
MADCTL_ML = 0x10  # LCD Refresh Bottom to Top
MADCTL_RGB = 0x00
MADCTL_BGR = 0x08
MADCTL_MH = 0x04  # LCD Refresh Right to Left

class ILI9341:
    def __init__(self, spi, cs, dc, rst, width=240, height=320, rotation=0):
        self.spi = spi
        self.cs = cs
        self.dc = dc
        self.rst = rst
        self.width = width
        self.height = height
        self.rotation = rotation % 4
        self.buffer = bytearray(width * height * 2)
        self.framebuf = framebuf.FrameBuffer(self.buffer, width, height, framebuf.RGB565)
        
        # Initialize display
        self.reset()
        self.init_display()
        
    def reset(self):
        self.rst(0)
        sleep_ms(10)
        self.rst(1)
        sleep_ms(120)
        
    def write_cmd(self, cmd):
        self.dc(0)
        self.cs(0)
        self.spi.write(bytes([cmd]))
        self.cs(1)
        
    def write_data(self, data):
        self.dc(1)
        self.cs(0)
        self.spi.write(data)
        self.cs(1)
        
    def init_display(self):
        self.write_cmd(ILI9341_SWRESET)
        sleep_ms(150)
        self.write_cmd(ILI9341_SLPOUT)
        sleep_ms(255)
        
        # Memory Data Access Control
        madctl = 0x00
        if self.rotation == 0:   # Portrait
            madctl = MADCTL_MX | MADCTL_BGR
        elif self.rotation == 1: # Landscape
            madctl = MADCTL_MV | MADCTL_MX | MADCTL_BGR
        elif self.rotation == 2: # Inverted Portrait
            madctl = MADCTL_MY | MADCTL_BGR
        elif self.rotation == 3: # Inverted Landscape
            madctl = MADCTL_MV | MADCTL_MY | MADCTL_BGR
        self.write_cmd(ILI9341_MADCTL)
        self.write_data(bytes([madctl]))
        
        # Pixel Format Set
        self.write_cmd(ILI9341_COLMOD)
        self.write_data(bytes([0x55]))  # 16-bit color
        
        self.write_cmd(ILI9341_DISPON)
        sleep_ms(10)
        
    def set_window(self, x0, y0, x1, y1):
        self.write_cmd(ILI9341_CASET)
        self.write_data(bytes([x0 >> 8, x0 & 0xFF, x1 >> 8, x1 & 0xFF]))
        self.write_cmd(ILI9341_PASET)
        self.write_data(bytes([y0 >> 8, y0 & 0xFF, y1 >> 8, y1 & 0xFF]))
        self.write_cmd(ILI9341_RAMWR)
        
    def fill(self, color):
        self.framebuf.fill(color)
        
    def text(self, string, x, y, color):
        self.framebuf.text(string, x, y, color)
        
    def vline(self, x, y, h, color):
        self.framebuf.vline(x, y, h, color)
        
    def show(self):
        self.set_window(0, 0, self.width - 1, self.height - 1)
        self.dc(1)
        self.cs(0)
        self.spi.write(self.buffer)
        self.cs(1)

def color565(r, g, b):
    return ((r & 0xF8) << 8) | ((g & 0xFC) << 3) | (b >> 3)

File: ili9341.c
#include "ili9341.h"
#include "pico/stdlib.h"
#include "hardware/spi.h"

// 8x8 font (subset, printable ASCII)
static const uint8_t font_8x8[95][8] = {
    {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, // Space
    {0x18, 0x18, 0x18, 0x18, 0x18, 0x00, 0x18, 0x00}, // !
    {0x36, 0x36, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}, // "
    {0x36, 0x36, 0x7F, 0x36, 0x7F, 0x36, 0x36, 0x00}, // #
    {0x0C, 0x3E, 0x03, 0x1E, 0x30, 0x1F, 0x0C, 0x00}, // $
    // Add more characters as needed
    {0x00, 0x00, 0x18, 0x18, 0x00, 0x18, 0x18, 0x00}, // :
    // Full font table omitted for brevity
};

static uint cs_pin, dc_pin, rst_pin;
static uint16_t buffer[320 * 240];

void ili9341_init(uint cs, uint dc, uint rst) {
    cs_pin = cs;
    dc_pin = dc;
    rst_pin = rst;
    gpio_init(cs_pin);
    gpio_init(dc_pin);
    gpio_init(rst_pin);
    gpio_set_dir(cs_pin, GPIO_OUT);
    gpio_set_dir(dc_pin, GPIO_OUT);
    gpio_set_dir(rst_pin, GPIO_OUT);
    gpio_put(cs_pin, 1);
    
    ili9341_reset();
    ili9341_write_cmd(ILI9341_SWRESET);
    sleep_ms(150);
    ili9341_write_cmd(ILI9341_SLPOUT);
    sleep_ms(255);
    
    ili9341_write_cmd(ILI9341_MADCTL);
    uint8_t madctl = MADCTL_MV | MADCTL_BGR;
    ili9341_write_data(&madctl, 1);
    
    ili9341_write_cmd(ILI9341_COLMOD);
    uint8_t colmod = 0x55;
    ili9341_write_data(&colmod, 1);
    
    ili9341_write_cmd(ILI9341_DISPON);
    sleep_ms(10);
}

void ili9341_reset() {
    gpio_put(rst_pin, 0);
    sleep_ms(10);
    gpio_put(rst_pin, 1);
    sleep_ms(120);
}

void ili9341_write_cmd(uint8_t cmd) {
    gpio_put(dc_pin, 0);
    gpio_put(cs_pin, 0);
    spi_write_blocking(SPI_PORT, &cmd, 1);
    gpio_put(cs_pin, 1);
}

void ili9341_write_data(uint8_t *data, size_t len) {
    gpio_put(dc_pin, 1);
    gpio_put(cs_pin, 0);
    spi_write_blocking(SPI_PORT, data, len);
    gpio_put(cs_pin, 1);
}

void ili9341_set_window(uint16_t x0, uint16_t y0, uint16_t x1, uint16_t y1) {
    ili9341_write_cmd(ILI9341_CASET);
    uint8_t data[4] = {x0 >> 8, x0 & 0xFF, x1 >> 8, x1 & 0xFF};
    ili9341_write_data(data, 4);
    ili9341_write_cmd(ILI9341_PASET);
    data[0] = y0 >> 8; data[1] = y0 & 0xFF;
    data[2] = y1 >> 8; data[3] = y1 & 0xFF;
    ili9341_write_data(data, 4);
    ili9341_write_cmd(ILI9341_RAMWR);
}

void ili9341_fill(uint16_t color) {
    for (int i = 0; i < 320 * 240; i++) buffer[i] = color;
}

void ili9341_text(const char *str, uint16_t x, uint16_t y, uint16_t color) {
    while (*str) {
        if (*str >= 32 && *str <= 126) {
            const uint8_t *glyph = font_8x8[*str - 32];
            for (int row = 0; row < 8; row++) {
                for (int col = 0; col < 8; col++) {
                    if (glyph[row] & (1 << (7 - col))) {
                        if (x + col < 320 && y + row < 240) {
                            buffer[(y + row) * 320 + (x + col)] = color;
                        }
                    }
                }
            }
        }
        x += 8;
        str++;
    }
}

void ili9341_vline(uint16_t x, uint16_t y, uint16_t h, uint16_t color) {
    if (x >= 320 || y >= 240) return;
    if (y + h > 240) h = 240 - y;
    for (int i = y; i < y + h; i++) {
        buffer[i * 320 + x] = color;
    }
}

void ili9341_fill_rect(uint16_t x, uint16_t y, uint16_t w, uint16_t h, uint16_t color) {
    if (x >= 320 || y >= 240) return;
    if (x + w > 320) w = 320 - x;
    if (y + h > 240) h = 240 - y;
    for (int i = y; i < y + h; i++) {
        for (int j = x; j < x + w; j++) {
            buffer[i * 320 + j] = color;
        }
    }
}

void ili9341_rect(uint16_t x, uint16_t y, uint16_t w, uint16_t h, uint16_t color) {
    ili9341_vline(x, y, h, color);
    ili9341_vline(x + w - 1, y, h, color);
    for (int i = y; i < y + h; i++) {
        buffer[i * 320 + x] = color;
        buffer[i * 320 + (x + w - 1)] = color;
    }
}

void ili9341_show() {
    ili9341_set_window(0, 0, 319, 239);
    gpio_put(dc_pin, 1);
    gpio_put(cs_pin, 0);
    spi_write_blocking(SPI_PORT, (uint8_t *)buffer, 320 * 240 * 2);
    gpio_put(cs_pin, 1);
}